### PR TITLE
feat: Disruption Diagram - Backend

### PR DESCRIPTION
**Pull request by jzimbel-mbta**
**Asana task**: [Write the serialization for the line map](https://app.asana.com/0/1185117109217413/1204736316281209/f)

The code is split into 5 modules.
I've ordered them below by how the logic generally flows, from public interface down to detailed implementation.

| Module | What's in it |
| - | - |
| `DisruptionDiagram` | The public interface of all the disruption diagram code. |
| `DisruptionDiagram.Model` | Contains high-level logic reflecting [Betsy's Miro flowchart](https://miro.com/app/board/uXjVP2Hgi18=/). |
| `DisruptionDiagram.Validator` | Logic to check that we can build a diagram from this localized alert. |
| `DisruptionDiagram.Builder` | Structure used to build the final diagram from a localized alert.
| `DisruptionDiagram.Label` | Logic to determine labels for omitted segments and destination-arrow end slots. |

Thoroughly tested! I needed to add some helpers to make writing test scenarios more manageable—those are in `test/support/`. All tests are against `DisruptionDiagram.serialize/1`, since it's really the only public interface of all of this code.

- [x] Tests added?

### CODE REVIEW

File: lib/screens/alerts/alert.ex
Comment (jzimbel-mbta) [reply lib/screens/alerts/alert.ex:580]: I think it does, sort of, in the same sense as the Green Line. But we don't have any screens along it, and afaik there are no current plans to add any, so the Screens code generally ignores it.

(We do have a pre-fare screen at Ashmont now, though, so maybe we should start paying more attention to it...)

File: lib/screens/v2/disruption_diagram/model.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/model.ex:230]: Yeah, it's a very simple use of metaprogramming. It's a way to generate a bunch of repetitive function clauses when the function essentially does a map lookup.

It generates:
```ex
defp baseline_slots(2), do: 10
defp baseline_slots(3), do: 10
defp baseline_slots(4), do: 12
...
```

The end result is functionally about the same as this:
```ex
@closure_to_baseline %{2 => 10, 3 => 10, 4 => 12, 5 => 12, 6 => 14, 7 => 14, 8 => 14}

defp baseline_slots(closure) do
  Map.fetch!(@closure_to_baseline, closure)
```

except it has a little more opportunity to be compile-time optimized, and if we call the function on a value that's not in the mapping, a `FunctionClauseError` is raised instead of a `KeyError`.

Overall... not really necessary, but I find that it feels a little bit more readable than the map-lookup version, once you get used to the `unquote` stuff.

File: lib/screens/alerts/alert.ex
Comment (PaulJKim) [reply lib/screens/alerts/alert.ex:580]: Yeah, I had Ashmont in mind and wasn't sure if any Mattapan-related content could appear on there., but seems like that's not a thing at the moment

File: lib/screens/v2/disruption_diagram/model.ex
Comment (PaulJKim) [reply lib/screens/v2/disruption_diagram/model.ex:230]: ohhh I see. That is pretty neat! I agree it's a bit more readable, as long as you understand the meta-programming basics

File: lib/screens/v2/widget_instance/reconstructed_alert.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/widget_instance/reconstructed_alert.ex:944]: Oh, I just remembered what this was! While we were testing the code out against some "real" alerts in dev-green, we needed to ignore a few alerts that others had created. This shouldn't have been checked in, I'll restore it.

Comment (jzimbel-mbta) [reply lib/screens/v2/widget_instance/reconstructed_alert.ex:944]: Actually, this code exactly matches what's on the target branch (`prefare-alerts`) right now: https://github.com/mbta/screens/blob/prefare-alerts/lib/screens/v2/widget_instance/reconstructed_alert.ex#L926

I'm not sure why it's being shown in the diff... I'll try merging `prefare-alerts` into this branch, but might just need to ignore this bit.

Comment (jzimbel-mbta) [reply lib/screens/v2/widget_instance/reconstructed_alert.ex]: Ok merging cleared it from the diff! Git still mystifies me sometimes... 🤷

**Review by hannahpurcell — COMMENTED** [review]
Phew, definitely a big task!! Pausing my first pass here, because I think I see a bug that might affect the logic I'm trying to review?

This RL alert only affects the Braintree branch. The current screen is at Porter, and the returned `routes` contain both the Ashmont and Braintree ids. I think only the Braintree id should be returned here.

Screenshot 2023-09-13 at 11 15 45 AM

Screenshot 2023-09-13 at 11 16 07 AM

**Comment by jzimbel-mbta** [comment]
Weird, I'm almost positive I had at least a handful of unit tests that covered Red Line branch cases. I'll take a look in a bit, thanks for the review!

File: lib/screens/v2/disruption_diagram/model.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/model.ex:41]: Sorry, this turned into a novel...

<details>
<summary><h3><code>rescue</code> / implicit try</h2></summary>

---
`rescue` is basically the `catch` clause of a try/catch. For some reason in elixir they call it `try`/`rescue` instead, but it works exactly the same—it catches any raised exceptions. If the try/catch is at the top level of the function, you can make it "implicit", which is what I did here. ([docs](https://hexdocs.pm/elixir/Kernel.html#def/2-rescue-catch-after-else))
If this code feels too "clever" I'd be happy to make the `try` expression explicit instead!

---

</details>

<details>
<summary><h3><code>with</code> without <code>else</code></h3></summary>

---

When a `with` doesn't have an `else` block, it just immediately returns anything that doesn't match one of the arrow clauses. Unlike the `->`'s in a `case` expression, a left-arrow clause `<-` doesn't raise an exception when a match fails. This is true in `for` expressions as well!

This:
```ex
with {:ok, data} <- fetch_data(),
     {:ok, entity} <- parse_data(data),
     {:ok, related_entity} <- Map.fetch(relation, entity.id) do
  {:ok, entity, related_entity}
end
```
is equivalent to this:
```ex
with {:ok, data} <- fetch_data(),
     {:ok, entity} <- parse_data(data),
     {:ok, related_entity} <- Map.fetch(relation, entity.id) do
  {:ok, entity, related_entity}
else
  non_matching_value -> non_matching_value
end
```
The first function call that returns something other than `{:ok, _some_value}` will make the `with` immediately return that value.
For example, if `fetch_data` succeeds but `parse_data` returns `{:error, some_message}` then the whole `with` will return `{:error, some_message}`.

You only need an `else` block if you need to modify some non-happy-path values to something else before returning. ([docs](https://hexdocs.pm/elixir/Kernel.SpecialForms.html#with/1-else-clauses))

If all of the called functions return something of the same shape for their failure cases, e.g. `{:error, message}`, then you don't really need an `else` block in your `with` expression. You can just return the error value as-is.

---

</details>

So basically, I was able to avoid `else` clauses on my `with`s because all of the "graceful failure" cases return the same `{:error, message}` shape which falls through straight to the top. That case is handled [here](https://github.com/mbta/screens/pull/1828/files#diff-14e0fcdf7283a1020ce87a425d3445b064821f56e9ee5034235bd3d5d7d73e4fR892) and documented in the [spec](https://github.com/mbta/screens/blob/d842bf7d8a7056bfdae16dfba771b34198cdc098/lib/screens/v2/disruption_diagram/model.ex#L31).

The `rescue` is there to separately deal with any _unexpected_ failures that result in an exception being raised, e.g. failing to look up a stop name [here](https://github.com/mbta/screens/blob/d842bf7d8a7056bfdae16dfba771b34198cdc098/lib/screens/v2/disruption_diagram/builder.ex#L117).

I wanted to make sure that **no matter what, the disruption diagram code can't cause the whole request to fail**, since it's an optional "nice-to-have" part of the alert widget.

File: lib/screens/v2/disruption_diagram/model.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/model.ex:32]: Currently, it's not possible in the one place this is called.
I think it's ok for us to fail here if that happens, because such a case should be easy to catch in local development—I can't imagine a scenario where it would happen inconsistently, since we would have to conditionally pass a whole different map shape to this function for that to happen.

File: lib/screens/v2/disruption_diagram/validator.ex
Comment (hannahpurcell) [reply lib/screens/v2/disruption_diagram/validator.ex:31]: 🤔 Hm, actually. 

Well we only fetch "relevant alerts" for pre-fare screens, so the alerts are only ever subway or light rail. This function only runs for a shuttle / suspension, which is line-based. So the informed entities of the alert should only have line-relevant routes. (There won't be lingering bus-related stops in that entities list.)

But totally fine to leave it in if that helps add security to this function.

File: lib/screens/v2/disruption_diagram/model.ex
Comment (hannahpurcell) [reply lib/screens/v2/disruption_diagram/model.ex:41]: Oh my! Waaaaay more detail than expected, but definitely thoughtful and makes all the sense. Thanks for the resources where I can learn more if needed. But sounds good to me

**Comment by hannahpurcell** [comment]
Not actively reviewing, but popped in to borrow your branch to build images for Transpo Camp. Seems like this alert won't build a diagram? Not sure why, but been a while since I've thought about disruption diagrams....

Screenshot 2023-10-05 at 3 51 05 PM

**Comment by jzimbel-mbta** [comment]
Re: the first problem alert on the RL Braintree branch, _<sup>(why doesn't GitHub let you create threaded replies on top-level PR comments? 😑)</sup>_ it's rendering correctly after I merged in latest `prefare-alerts`! The issue was some out-of-date frontend logic.

Screenshot 2023-10-11 at 11-54-58 Screens

**Comment by jzimbel-mbta** [comment]
The second one is a bug! We should be able to produce a diagram for that alert, at a Gov Ctr screen, since all stops are along the Green-C route. Looking into it.

image

File: lib/screens/v2/disruption_diagram/validator.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/validator.ex]: Moved to a new `Screens.Alerts.InformedEntity` module. I'm not sure if that's the best place for it within the namespace hierarchy, but it seemed like a thing that "belongs" to alerts so... 🤷

I only added two little functions to it, but I could imagine us moving more shared logic into it in the future.

**Comment by jzimbel-mbta** [comment]
Fixed the GL diagram by filtering to canonical route patterns!

<details>
<summary>Screenshot</summary>

image

</details>

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (hannahpurcell) [inline lib/screens/v2/disruption_diagram/builder.ex:551]: I feel a bit lost here. I get a vague sense of what is happening and the expected output, but am a bit thrown by the new_entries and why `first_disrupted_stop` only gets set once?

**Review by hannahpurcell — COMMENTED** [review]
The Back Bay DUP is not loving a change to how its route_pattern logic for alerts is working, though there are no OL alerts affecting it at the moment. I'm not seeing this issue in `prefare-alerts` or master, just the feature branch

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (hannahpurcell) [inline lib/screens/v2/disruption_diagram/builder.ex:192]: How do I interpret the first part of this line?

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (hannahpurcell) [inline lib/screens/v2/disruption_diagram/builder.ex:803]: OH! I think I understand what left_end and right_end are for -- they have all the stops not currently going into the rest of the diagram? If that's the case, can that be a comment in the top of this file where left_end / right_end are defined?

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (hannahpurcell) [inline lib/screens/v2/disruption_diagram/builder.ex:479]: This one could use a comment, too

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (hannahpurcell) [inline lib/screens/v2/disruption_diagram/builder.ex:753]: Another comment here would be good

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (hannahpurcell) [inline lib/screens/v2/disruption_diagram/builder.ex:902]: What's the purpose of this separate def of `closure_indices` since there's only one pattern here, and no transformation happens?

**Review by hannahpurcell — COMMENTED** [review]
The one file I haven't fully wrapped my head around is the `builder`, but I think I'll pick it up through applied work on it. Also @cmaddox5 is helping give it a secondary look, so as long as enough folks on our team can understand it and answer questions about it, then we should be good.

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (cmaddox5) [inline lib/screens/v2/disruption_diagram/builder.ex:1]: Gave this file a second pair of eyes and I don't have anything to add. The iffy parts for me already had comments from Hannah. Any work on this in the future will definitely have a steep learning curve, but the logic is sound. Great work!

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/builder.ex:469]: > Is it true that when get_builder_data returns a branch, that means both ends of the sequence are terminals?

Yes, at least at the start.

This function serves to put in arrow slots where we are 100% sure it's a GL/RL trunk alert with the home stop also on the trunk. In those cases, we never want the diagram to go beyond a branch point (Lechmere, Kenmore, JFK), because we wouldn't know which branch to choose, so the relevant ends have to use arrows.

In all other cases, we simply mark the first/last stops as terminals. In subsequent logic they may yet be replaced with arrow slots.

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/builder.ex:191]: I can do that, but `try_omit` itself calls a whole bunch of helper and sub-(sub-(sub-))helper functions. Would you like me to just move `try_omit` up, or all of the omission-related private functions? `try_omit`, `do_omit`, `try_alternate_omit`, `try_side_omit`, `try_split_omit`, `find_safe_segment`, `select_split_omission_indices`

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/builder.ex:902]: I originally separated each of these `_indices` funcs into 2 layers in order to make the code more explicit during development, but at this point they can be cleaned up. Especially since I worked with Betsy to adjust things so that the ends have lowest precedence instead of current location.

`<region>_ideal_indices` gives the indices that the region would contain if none were taken by a higher-precedence region, and `<region>_indices` gives the actual indices that it gets after other regions take away their respective pieces.

Since the closure region has the highest precedence and can't have any stops taken from it, these are one in the same.

File: test/support/disruption_diagram_localized_alert.ex
Comment (jzimbel-mbta) [reply test/support/disruption_diagram_localized_alert.ex:1]: 🎗️ Noted

File: lib/screens/v2/disruption_diagram/model.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/model.ex:41]: I've added a little glossary of terms to the moduledoc, let me know if it's helpful.

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/builder.ex:568]: `Vector.foldl` is the same as `Enum.reduce`, if that helps. You start with an initial accumulator, `meta_without_indices` in this case, and "reduce" the list down to a single value.

At a high level this is re-analyzing `sequence` to find the new indices of important slots, after `sequence` has been changed in some way.

For each slot in `sequence`, moving from left to right, if it's a `StopSlot` we check if that stop is disrupted or the home stop.

If it's disrupted:
- we try setting `first_disrupted_stop` if it wasn't already set. If it was, then that means this isn't the first one we've found!
- we always set `last_disrupted_stop`, because we want to keep updating that value until we reach the actual last disrupted stop.

If it's the home stop then we set the `home_stop`. There's only one of those.

I wrote the code this way to have the update happen in a single pass of `sequence`, but I agree it's not the clearest. Would this approach be better? It shouldn't have much of a performance impact.
```ex
# This code would replace the
# `new_metadata = builder.sequence |> Vector.with_index() |> Vector.foldl(...) |> then(...)`
# expression.

indexed_sequence = Vector.with_index(builder.sequence)

home_stop =
  Aja.Enum.find_value(indexed_sequence, fn
    {%StopSlot{home_stop?: true}, i} -> i
    _ -> false
  end)

first_disrupted_stop =
  Aja.Enum.find_value(indexed_sequence, fn
    {%StopSlot{disrupted?: true}, i} -> i
    _ -> false
  end)

last_disrupted_stop =
  indexed_sequence
  |> Vector.reverse()
  |> Aja.Enum.find_value(fn
    {%StopSlot{disrupted?: true}, i} -> i
    _ -> false
  end)

new_metadata =
  meta_without_indices
  |> Map.merge(%{
    home_stop: home_stop,
    first_disrupted_stop: first_disrupted_stop,
    last_disrupted_stop: last_disrupted_stop
  })
  |> then(&struct!(Metadata, &1))
```

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/builder.ex:198]: `closure = closure_indices(builder)` - binds the whole range to `closure`
`closure_first..closure_last//1 = closure` - binds the first and last values of the range to `closure_first` and `closure_last`, and asserts that the step of the range is 1

This is equivalent to pattern matches in `case` and such like:
```ex
case %{foo: :bar} do
  %{foo: foo_variable_binding} = map_containing_foo -> ...
end
```
which bind multiple variables to different parts/nesting levels of a compound value.

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/builder.ex]: Clarifying since it was buried in my last comment: I cleaned up this code and removed `closure_ideal_indices` and `gap_ideal_indices`.

**Comment by jzimbel-mbta** [comment]
> The Back Bay DUP is not loving a change to how its route_pattern logic for alerts is working

This should be fixed now, by [this commit](https://github.com/mbta/screens/pull/1828/commits/9461d49de032c88bcc214ba52bad1098de81dd46). My code previously defaulted to passing `filter[canonical]=false`, which to my surprise limits the results to only *non*-canonical route patterns. Definitely not what we want.

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (hannahpurcell) [reply lib/screens/v2/disruption_diagram/builder.ex:569]: I do think the first round is great for its conciseness, but I much much prefer this rewrite for its readability!! And if it doesn't impact performance, then I think this would be better in the long-run

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (hannahpurcell) [reply lib/screens/v2/disruption_diagram/builder.ex:198]: Ahhhh.... I see what you're saying and don't feel strongly about the rearrangement. Follow your heart!

**Review by hannahpurcell — APPROVED** [review]
Great! Gave a thumbs-up on two things you suggested. Once those have been applied, I think we're ready to go. Let's merge it and start bug-bashing!! 🐛

File: lib/screens/v2/disruption_diagram/builder.ex
Comment (jzimbel-mbta) [reply lib/screens/v2/disruption_diagram/builder.ex]: I ended up moving `get_important_indices` down closer to where it's used! It was in the wrong place anyway, up among all the public functions. Now all public functions come first and are followed by the private helpers.

**Comment by jzimbel-mbta** [comment]
I'm not sure why CI stopped running on my commits, but I manually checked and all unit tests are passing, plus Credo and Dialyzer have no complaints. Merging!

File: test/support/disruption_diagram_localized_alert.ex
Comment (jzimbel-mbta) [reply test/support/disruption_diagram_localized_alert.ex:1]: ✅

**Pull request merged**
